integration_tests/src/lib.rs

```diff
@@ -24,47 +24,89 @@ pub struct TaskParameters {
     pub collector_auth_token: AuthenticationToken,
 }
 
-/// Components of DAP endpoints for a leader and helper aggregator. By default, the scheme is
-/// assumed to be `http:`, and the port number is assumed to be 8080.
+/// Components of one aggregator's DAP endpoint. The scheme is assumed to always be `http:`.
+#[derive(Debug)]
+pub enum AggregatorEndpointFragments {
+    /// The aggregator is in a virtual network, (a Docker network or a Kind cluster) so different
+    /// URLs must be used depending on whether it is accessed from within the same virtual network
+    /// or via a port forward on localhost. It is assumed that the port will always be 8080 within
+    /// the virtual network. The port number of forwarded ports will be supplied later.
+    VirtualNetwork { host: String, path: String },
+    /// The aggregator is running on localhost. No port forwarding is involved, so the same URL is
+    /// used in all circumstances. The port number will be supplied later.
+    Localhost { path: String },
+}
+
+impl AggregatorEndpointFragments {
+    /// Provides the URL for the aggregator's endpoint from the perspective of the host. If the
+    /// aggregator is in a virtual network, this will use a port forward, with the provided port
+    /// number. If the aggregator itself is running on the host, it will use the provided port
+    /// number as well.
+    pub fn endpoint_for_host(&self, port: u16) -> Url {
+        let path = match self {
+            AggregatorEndpointFragments::VirtualNetwork { path, .. } => path,
+            AggregatorEndpointFragments::Localhost { path } => path,
+        };
+        Url::parse(&format!("http://127.0.0.1:{port}{path}")).unwrap()
```

Comment:
nit: should we use `http://localhost:{port}{path}` here just in case someone only has an IPv6 localhost? Though I notice that this was hard-coded to 127.0.0.1 in the existing code this refactors, so maybe we deliberately forced this to use IPv4 to avoid port collisions across the protocols...

Reply:
Using `localhost` leads to problems when connecting to Docker container ports, because the same exposed container port may be published at different ephemeral ports on 127.0.0.1 and ::1. When you combine that with our HTTP client doing happy eyeballs on both addresses for `localhost`, that resulted in wires getting crossed between different concurrently running tests. Sticking with 127.0.0.1 here and `Container::get_host_port_ipv4()` elsewhere keeps things straight.